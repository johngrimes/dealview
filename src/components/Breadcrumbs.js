import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import _ from 'lodash'

import './styles/Breadcrumbs.css'

class Breadcrumbs extends React.Component {
  render() {
    const breadcrumbLinks = _.map(this.props.breadcrumbs, (breadcrumb, i) => {
      return (
        <li className='breadcrumbs-item' key={i}>
          <Link className='breadcrumbs-link' to={breadcrumb.path}>
            {breadcrumb.display}
          </Link>
        </li>
      )
    })
    return (
      <div className='breadcrumbs'>
        <ul className='breadcrumbs-list'>
          {breadcrumbLinks}
        </ul>
      </div>
    )
  }
}

Breadcrumbs.propTypes = {
  breadcrumbs: PropTypes.arrayOf(
    PropTypes.shape({
      display: PropTypes.string,
      path: PropTypes.string,
    })
  ),
}

export default Breadcrumbs
