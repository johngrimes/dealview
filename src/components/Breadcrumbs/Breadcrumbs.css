.breadcrumbs-list {
  list-style-type: none;
  margin: 0;
  padding: 8px 16px;
  background-color: #eee;
}

.breadcrumbs-item {
  display: inline-block;
  margin-right: 10px;
}

.breadcrumbs-item::after {
  margin-left: 5px;
  content: "\2192";
}

.breadcrumbs-item:last-child::after {
  margin: 0;
  content: "";
}

.breadcrumbs-link {
  color: black;
  text-decoration: none;
  font-size: 14px;
}

.breadcrumbs-link:hover {
  text-decoration: underline;
}
