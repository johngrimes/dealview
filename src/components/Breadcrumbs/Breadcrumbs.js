import React from 'react'
import { Link } from 'react-router'

import './Breadcrumbs.css'

export type Breadcrumb = {
  display: string,
  path: string
}

type Component = {
  breadcrumb?: Breadcrumb
}

export type Route = {
  component?: Component
}

// Recursive function that walks the supplied `routes` object, calling the
// `breadcrumb` function on each component.
const getBreadcrumbsFromRoutes = (routes: Array<Route>): Array<Breadcrumb> => {
  const tail = routes.slice(1)
  const head = routes[0]

  if (!head) {
    return []
  } else if (head.component && head.component.breadcrumb) {
    const breadcrumb = head.component.breadcrumb()
    const newBreadcrumbs = getBreadcrumbsFromRoutes(tail)
    newBreadcrumbs.push(breadcrumb)
    return newBreadcrumbs
  }
  return getBreadcrumbsFromRoutes(tail)
}

// Takes an array of Breadcrumb objects and returns an array of links, each
// wrapped in a list item.
const breadcrumbLinks = (breadcrumbs: Array<Breadcrumb>): Array<React$Element> => {
  const breadcrumbsCopy = breadcrumbs.slice()
  const breadcrumb = breadcrumbsCopy.pop()

  if (breadcrumb) {
    const breadcrumbLink = <li className='breadcrumbs-item' key={breadcrumb.path}>
                             <Link className='breadcrumbs-link' to={breadcrumb.path}>
                               {breadcrumb.display}
                             </Link>
                           </li>
    const newBreadcrumbLinks = breadcrumbLinks(breadcrumbsCopy)

    newBreadcrumbLinks.unshift(breadcrumbLink)

    return newBreadcrumbLinks
  }

  return []
}

type Props = {
  routes: Array<Route>
}

// A Component that displays a breadcrumb trail, based on a set of routes
// passed in via the `routes` prop.
class Breadcrumbs extends React.Component {
  props: Props

  constructor(props) {
    super(props)
    this.state = { breadcrumbs: getBreadcrumbsFromRoutes(props.routes) }
  }

  render() {
    return (
      <div className='breadcrumbs'>
        <ul className='breadcrumbs-list'>
          {breadcrumbLinks(this.state.breadcrumbs)}
        </ul>
      </div>
    )
  }
}

export default Breadcrumbs
