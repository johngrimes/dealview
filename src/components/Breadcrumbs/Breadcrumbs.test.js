/* global expect */

import Breadcrumbs from './Breadcrumbs.js'

import React from 'react'
import { Link } from 'react-router-dom'
import { shallow } from 'enzyme'

describe('Breadcrumbs', () => {
  it('should render a breadcrumb trail', () => {
    const props = {
      breadcrumbs: [
        { display: 'Link 1', path: '/link1' },
        { display: 'Link 2', path: '/link2' }
      ]
    }
    const wrapper = shallow(<Breadcrumbs {...props} />)
    expect(wrapper.find(Link)).toHaveLength(2)
    expect(wrapper.findWhere(node => {
      return node.type() === Link &&
        node.children().text() === 'Link 1' &&
        node.prop('to') === '/link1'
    })).toHaveLength(1)
    expect(wrapper.findWhere(node => {
      return node.type() === Link &&
        node.children().text() === 'Link 2' &&
        node.prop('to') === '/link2'
    })).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })
})
