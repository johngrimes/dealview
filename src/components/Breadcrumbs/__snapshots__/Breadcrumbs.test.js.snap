exports[`Breadcrumbs should render a breadcrumb trail 1`] = `
<div
  className="breadcrumbs">
  <ul
    className="breadcrumbs-list">
    <li
      className="breadcrumbs-item">
      <Link
        className="breadcrumbs-link"
        replace={false}
        to="/link1">
        Link 1
      </Link>
    </li>
    <li
      className="breadcrumbs-item">
      <Link
        className="breadcrumbs-link"
        replace={false}
        to="/link2">
        Link 2
      </Link>
    </li>
  </ul>
</div>
`;
